man/socketmaster.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SOCKETMASTER" "1" "September 2015" "PandaStream" ""
.
.SH "NAME"
\fBsocketmaster\fR \- Zero downtime restart for your apps
.
.SH "SYNOPSIS"
\fBsocketmaster\fR [\-listen=URI] [\-command=PATH] [\-start=MILLIS] [\-syslog]
.
.SH "DESCRIPTION"
socketmaster is a generic solution to the restart problem: you want to restart your app but neither lose any active connection nor refuse new connections\.
.
.P
socketmaster opens the socket you want to listen on and passes it onto your program\. Because socketmaster is simple enough it doesn\'t need to be restarted and can keep your file\-descriptor indefinitely open\.
.
.TP
\fB\-command\fR=PATH
Defines the location of the program to execute and pass the file\-descriptor onto\. If you need to add arguments to the program, use \-\- before the command specific arguments or use a wrapper script\.
.
.IP
e\.g\. socketmaster \-command \fIcommand\fR
.
.TP
\fB\-listen\fR=ADDR
Depending on the number of times it\'s passed the child will get 0\-N sockets\.
.
.IP
Examples: :8080 fd://3 tcp://:8945 tcp4://localhost:8945 tcp6://localhost:8945 unix:///tmp/myapp\.sock
.
.TP
\fB\-start\fR=MILLIS
Determines how long socketmaster waits before signaling the old process\. If MILLIS is set to zero that feature is disabled\.
.
.TP
\fB\-syslog\fR
Tells socketmaster to log to syslog\.
.
.TP
\fB\-user\fR=LOGIN
Sets the child process\'s uid/gid to the given user (looked up in /etc/passwd)\. This command only works when socketmaster is run as root\.
.
.SH "HOW IT WORKS"
On start:
.
.IP "\(bu" 4
socketmaster open the \-listen port
.
.IP "\(bu" 4
socketmaster starts the \-command with the socket on fd 3 and LISTEN_FDS=1
.
.IP "\(bu" 4
as soon as all child processes are gone, socketmaster stops as well
.
.IP "" 0
.
.P
On SIGHUP:
.
.IP "\(bu" 4
socketmaster starts a new \-command
.
.IP "\(bu" 4
waits for \-start milliseconds
.
.IP "\(bu" 4
sends a SIGTERM to the other child processes
.
.IP "" 0
.
.P
If the new command is not loaded after \-start milliseconds the operation is aborted\. All SIGHUP happening at that time are ignored\.
.
.P
Your server is responsible for:
.
.IP "\(bu" 4
opening the socket passed on fd 3
.
.IP "\(bu" 4
not crashing
.
.IP "\(bu" 4
gracefully shutdown on SIGTERM (close the listener, wait for the child connections to close)
.
.IP "" 0
.
.SH "ENVIRONMENT"
The \fBLISTEN_*\fR environment variables follow the systemd socket activation documentation\. The child can use the "sd\-daemon\.h" library \fIhttp://0pointer\.de/blog/projects/socket\-activation\.html\fR\.
.
.TP
\fBLISTEN_FDS\fR
Number of file\-descriptors passed by socketmaster\. The child starts counting from 3\.
.
.TP
\fBLISTEN_PID\fR
The PID of the target child\. If the pids don\'t match the child should ignore the \fBLISTEN_FDS\fR environment variable\.
.
.SH "SIGNALS"
On SIGHUP, socketmaster starts a new child process, waits for \-start and sends a HUP to the old process\.
.
.P
SIGINT, SIGTERM and SIGQUIT are forwarded to the child processes\.
.
.SH "RELATED PROJECTS"
.
.IP "\(bu" 4
crank \fIhttps://github\.com/pusher/crank\fR is the child of socketmaster\. It supports controlled rolling restarts and dynamic config\.
.
.IP "\(bu" 4
einhorn \fIhttps://github\.com/stripe/einhorn\fR is a project with the same goals, a wider feature\-set and written in ruby\.
.
.IP "\(bu" 4
libancillary \fIhttp://www\.normalesup\.org/~george/comp/libancillary/\fR is a cross\-platform C library to work with file\-descriptors\.
.
.IP "\(bu" 4
portlisten \fIhttp://www\.mail\-archive\.com/kragen\-hacks@canonical\.org/msg00002\.html\fR another approach to port sharing trough UNIX sockets\.
.
.IP "\(bu" 4
systemd \fIhttp://freedesktop\.org/wiki/Software/systemd/\fR is an init system / process manager who supports socket activation and delegation\.
.
.IP "\(bu" 4
circus \fIhttp://circus\.readthedocs\.org/en/latest/\fR is a process manager with socket delegation\.
.
.IP "\(bu" 4
upstart \fIhttp://upstart\.ubuntu\.com/\fR is an init system / process manager\. See http://manpages\.ubuntu\.com/manpages/precise/en/man7/socket\-event\.7\.html
.
.IP "" 0
.
.SH "BUGS & CODE"
Bug reports and other contributions are welcome! The project\'s source and issue tracker is located at \fIhttps://github\.com/zimbatm/socketmaster/\fR
.
.SH "COPYRIGHT"
socketmaster is Copyright (C) 2012 PandaStream \fIhttp://pandastream\.com\fR socketmaster is Copyright (C) 2015 zimbatm \fIhttp://zimbatm\.com\fR
.
.P
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
.
.P
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software\.
.
.P
THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT\. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE\.


CHANGELOG.md

2.0.0 / 2015-09-06
==================

  * BREAK: Drop einhorn compatibility. They changed everything.
  * NEW: Adopt systemd socket-activation `LISTEN_FDS` and `LISTEN_PID`
         environment variables.
  * NEW: -listen can be called multiple times
  * NEW: -listen accepts the stdlib `:8080` format
  * NEW: socketmaster/slave interface for clients. Thanks @vivienschilis

1.0.57 / 2015-04-11
===================

  * CHANGE: HUP to TERM for shutting down clients. Thanks @qzaidi
  * NEW: allow to pass command-line arguments to the running script. Thanks @qzaidi
  * FIX: Avoid double [pid] tags on the syslog output
  * Doc typos and grammar, thanks @nikai3d and @patrickberkeley

1.0.40 / 2013-06-18
==================

  * Build socketmaster with go v1.1.1 - Fixes weird issues with shared fds.
  * Avoid double [pid] tags on the syslog output
  * Add a -user option to set the child process's uid/gid
  * Add a -syslog option to log to syslog
  * Fixed a race condition in the process set
  * Forward all child process output to the socketmaster logger
  * Fixed a race condition when two signals arrive at the same time
  * Allow to disable the start-wait

1.0.13 / 2012-11-12
===================

  * Prefix output with the [pid]
  * Set EINHORN_FDS to be eninhorn-compatible
  * Adding license and changelog
  * Add a note about other related projects

1.0.7 / 2012-11-11
==================

  * First release with just the basic features



socketmaster.go
package main

import (
	"flag"
	"fmt"
	"log"
	"log/syslog"
	"os"
	"os/exec"
	"os/signal"
	"os/user"
	"syscall"
	"time"
)

const PROGRAM_NAME = "socketmaster"

func handleSignals(processGroup *ProcessGroup, c <-chan os.Signal, startTime int) {
	for {
		signal := <-c // os.Signal
		syscallSignal := signal.(syscall.Signal)

		switch syscallSignal {
		case syscall.SIGHUP:
			process, err := processGroup.StartProcess()
			if err != nil {
				log.Printf("Could not start new process: %v\n", err)
			} else {
				if startTime > 0 {
					time.Sleep(time.Duration(startTime) * time.Millisecond)
				}

				// A possible improvement woud be to only swap the
				// process if the new child is still alive.
				processGroup.SignalAll(syscall.SIGTERM, process)
			}
		default:
			// Forward signal
			processGroup.SignalAll(signal, nil)
		}
	}
}

func main() {
	var (
		addr      string
		command   string
		err       error
		startTime int
		useSyslog bool
		username  string
	)

	flag.StringVar(&command, "command", "", "Program to start")
	flag.StringVar(&addr, "listen", "tcp://:8080", "Port on which to bind")
	flag.IntVar(&startTime, "start", 3000, "How long the new process takes to boot in millis")
	flag.BoolVar(&useSyslog, "syslog", false, "Log to syslog")
	flag.StringVar(&username, "user", "", "run the command as this user")
	flag.Parse()

	if useSyslog {
		stream, err := syslog.New(syslog.LOG_INFO, PROGRAM_NAME)
		if err != nil {
			panic(err)
		}
		log.SetFlags(0) // disables default timestamping
		log.SetOutput(stream)
		log.SetPrefix("")
	} else {
		log.SetFlags(log.Ldate | log.Ltime)
		log.SetOutput(os.Stderr)
		log.SetPrefix(fmt.Sprintf("%s[%d] ", PROGRAM_NAME, syscall.Getpid()))
	}

	if command == "" {
		log.Fatalln("Command path is mandatory")
	}

	commandPath, err := exec.LookPath(command)
	if err != nil {
		log.Fatalln("Could not find executable", err)
	}

	log.Println("Listening on", addr)
	sockfile, err := ListenFile(addr)
	if err != nil {
		log.Fatalln("Unable to open socket", err)
	}

	var targetUser *user.User
	if username != "" {
		targetUser, err = user.Lookup(username)
		if err != nil {
			log.Fatalln("Unable to find user", err)
		}
	}

	// Run the first process
	processGroup := MakeProcessGroup(commandPath, sockfile, targetUser)
	_, err = processGroup.StartProcess()
	if err != nil {
		log.Fatalln("Could not start process", err)
	}

	// Monitoring the processes
	c := make(chan os.Signal)
	signal.Notify(c, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGHUP)
	go handleSignals(processGroup, c, startTime)

	// TODO: Full restart on USR2. Make sure the listener file is not set to SO_CLOEXEC
	// TODO: Restart processes if they die

	// For now, exit if no processes are left
	processGroup.WaitAll()
}
